File: app/src/main/java/com/chess/cryptobot/view/notification/NotificationBuilder.java
package com.chess.cryptobot.view.notification;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import com.chess.cryptobot.R;
import com.chess.cryptobot.view.MainActivity;

public class NotificationBuilder {
    private final Context context;
    private int notificationId;
    private String channelId;
    private String channelName;
    private String text;
    private String extraFlag;
    private int importance;
    private String title;
    private Integer color;
    private final NotificationManager notificationManager;

    public NotificationBuilder(Context context) {
        this.context = context;
        notificationManager = context.getSystemService(NotificationManager.class);
    }

    public NotificationBuilder setNotificationId(int id) {
        this.notificationId = id;
        return this;
    }

    public NotificationBuilder setChannelId(String channelId) {
        this.channelId = channelId;
        return this;
    }

    public NotificationBuilder setNotificationText(String text) {
        this.text = text;
        return this;
    }

    public NotificationBuilder setExtraFlag(String flag) {
        this.extraFlag = flag;
        return this;
    }

    public NotificationBuilder setChannelName(String channelName) {
        this.channelName = channelName;
        return this;
    }

    public NotificationBuilder setImportance(int importance) {
        this.importance = importance;
        return this;
    }

    public NotificationBuilder setTitle(String title) {
        this.title = title;
        return this;
    }

    public NotificationBuilder setColor(Integer color) {
        this.color = color;
        return this;
    }

    public void buildAndNotify() {
        notificationManager.notify(notificationId, build());
    }

    public Notification build() {

        createNotificationChannelIfNotExist(notificationManager);
        return buildNotification();

    }

    private void createNotificationChannelIfNotExist(NotificationManager notificationManager) {
        NotificationChannel channel = notificationManager.getNotificationChannel(channelId);
        if (channel == null) {
            channel = new NotificationChannel(channelId, channelName, importance);
            channel.enableVibration(false);
            channel.enableLights(false);
            notificationManager.createNotificationChannel(channel);
        }
    }

    private Notification buildNotification() {
        PendingIntent pendingIntent = getPendingIntent();

        Notification.Builder builder = new Notification.Builder(context, channelId)
                .setSmallIcon(R.drawable.round_monetization_on_24)
                .setContentTitle(title)
                .setStyle(new Notification.BigTextStyle().bigText(text))
                .setAutoCancel(true)
                .setColorized(true)
                .setContentIntent(pendingIntent)
                .setCategory(Notification.CATEGORY_SERVICE);
        if (this.color != null) {
            builder.setColorized(true)
                    .setColor(context.getResources().getColor(R.color.colorPrimary, null));
        }

        return builder.build();
    }

    private PendingIntent getPendingIntent() {
        Intent intent = new Intent(context, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        if (extraFlag != null) intent.putExtra(extraFlag, true);
        int uniqueInt = (int) (System.currentTimeMillis() & 0xfffffff);
        return PendingIntent.getActivity(context, uniqueInt, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }
}

File: app/src/main/java/com/chess/cryptobot/view/notification/NotificationBuilder.kt
package com.chess.cryptobot.view.notification

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import com.chess.cryptobot.R
import com.chess.cryptobot.view.MainActivity

class NotificationBuilder(private val context: Context) {
    private var notificationId = 0
    private var channelId: String? = null
    private var channelName: String? = null
    private var text: String? = null
    private var extraFlag: String? = null
    private var importance = 0
    private var title: String? = null
    private var color: Int? = null
    private val notificationManager: NotificationManager = context.getSystemService(NotificationManager::class.java)

    fun setNotificationId(id: Int): NotificationBuilder {
        notificationId = id
        return this
    }

    fun setChannelId(channelId: String?): NotificationBuilder {
        this.channelId = channelId
        return this
    }

    fun setNotificationText(text: String?): NotificationBuilder {
        this.text = text
        return this
    }

    fun setExtraFlag(flag: String?): NotificationBuilder {
        extraFlag = flag
        return this
    }

    fun setChannelName(channelName: String?): NotificationBuilder {
        this.channelName = channelName
        return this
    }

    fun setImportance(importance: Int): NotificationBuilder {
        this.importance = importance
        return this
    }

    fun setTitle(title: String?): NotificationBuilder {
        this.title = title
        return this
    }

    fun setColor(color: Int?): NotificationBuilder {
        this.color = color
        return this
    }

    fun buildAndNotify() {
        notificationManager.notify(notificationId, build())
    }

    fun build(): Notification {
        createNotificationChannelIfNotExist(notificationManager)
        return buildNotification()
    }

    private fun createNotificationChannelIfNotExist(notificationManager: NotificationManager) {
        var channel = notificationManager.getNotificationChannel(channelId)
        if (channel == null) {
            channel = NotificationChannel(channelId, channelName, importance)
            channel.enableVibration(false)
            channel.enableLights(false)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun buildNotification(): Notification {
        val pendingIntent = pendingIntent
        val builder = Notification.Builder(context, channelId)
                .setSmallIcon(R.drawable.round_monetization_on_24)
                .setContentTitle(title)
                .setStyle(Notification.BigTextStyle().bigText(text))
                .setAutoCancel(true)
                .setColorized(true)
                .setContentIntent(pendingIntent)
                .setCategory(Notification.CATEGORY_SERVICE)
        if (color != null) {
            builder.setColorized(true)
                    .setColor(context.resources.getColor(R.color.colorPrimary, null))
        }
        return builder.build()
    }

    private val pendingIntent: PendingIntent
        get() {
            val intent = Intent(context, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
            if (extraFlag != null) intent.putExtra(extraFlag, true)
            val uniqueInt = (System.currentTimeMillis() and 0xfffffff).toInt()
            return PendingIntent.getActivity(context, uniqueInt, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        }

}